### PR TITLE
HIVE-26244: Implementing locking for concurrent ctas

### PR DESCRIPTION
**Pull request by simhadri-g**
### What changes were proposed in this pull request?

1. Address the issue with concurrent CTAS operations when creating a table with the same name.
2. Introduce a new dataOperationType for CTAS(t).
3. Change the lock that is taken by the CTAS operation for transactional tables from DDL_NO_LOCK to EXCL_WRITE lock.
4. Check for entries in TXN_COMPONETS table when a concurrent CTAS operation (to create the same table) is blocked and fail-early the concurrent CTAS operation as it is unnecessary.

### Why are the changes needed?

1. Currently, the CTAS operation does not acquire any lock(ie DDL_NO_LOCK). 
2. Let us say that there are 2 concurrent CTAS operations on the same target table. This will result in a race to determine who commits first. The query that commits first will succeed whereas the query that is yet to commit will fail with the table already exists exception. This will result in an unnecessary overhead of cleaning up any data written by the failed query and a significant amount of unnecessary move tasks.

With this PR, CTAS operation for transactional tables will acquire EXCL_WRITE lock and fail-early any other concurrent ctas query that tries to create the same table.

### Does this PR introduce _any_ user-facing change?

New conf is introduced  hive.acid.check.for.concurrent.ctas.enabled
### How was this patch tested?

Unit tests and Q tests.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 1]
```diff
@@ -13900,7 +13901,11 @@ private void addDbAndTabToOutputs(String[] qualifiedTabName, TableType type,
     for(Map.Entry<String,String> serdeMap : storageFormat.getSerdeProps().entrySet()){
       t.setSerdeParam(serdeMap.getKey(), serdeMap.getValue());
     }
-    outputs.add(new WriteEntity(t, WriteEntity.WriteType.DDL_NO_LOCK));
+    if (tblProps.get("created_with_ctas") == "true") {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13904]: is there any constant for "created_with_ctas"?

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java [hunk 2]
```diff
@@ -18,24 +18,13 @@
 package org.apache.hadoop.hive.ql.lockmgr;
 
 import org.apache.commons.lang3.StringUtils;
-import org.apache.hadoop.fs.FileStatus;
-import org.apache.hadoop.fs.FileSystem;
-import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.*;
```
Comment (deniskuzZ) [inline ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:21]: please remove wildcard imports

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java [hunk 3]
```diff
@@ -18,24 +18,13 @@
...
-import org.apache.hadoop.hive.metastore.api.DataOperationType;
-import org.apache.hadoop.hive.metastore.api.LockState;
-import org.apache.hadoop.hive.metastore.api.LockType;
-import org.apache.hadoop.hive.metastore.api.ShowLocksRequest;
-import org.apache.hadoop.hive.metastore.api.ShowLocksResponse;
-import org.apache.hadoop.hive.metastore.api.ShowLocksResponseElement;
-import org.apache.hadoop.hive.metastore.api.TxnType;
-import org.apache.hadoop.hive.metastore.api.CommitTxnRequest;
+import org.apache.hadoop.hive.metastore.Warehouse;
+import org.apache.hadoop.hive.metastore.api.*;
```
Comment (deniskuzZ) [inline ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:27]: please remove wildcard imports

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableOperation.java [hunk 4]
```diff
@@ -98,8 +98,11 @@ public int execute() throws HiveException {
       }
       createTableNonReplaceMode(tbl);
     }
-
-    DDLUtils.addIfAbsentByName(new WriteEntity(tbl, WriteEntity.WriteType.DDL_NO_LOCK), context);
+    if (context.getQueryState().getCommandType().equals("CREATETABLE_AS_SELECT")) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableOperation.java:101]: should we generalize this to ACID create statements, not only CTAS?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 5]
```diff
@@ -5283,6 +5284,39 @@ is performed on that db (e.g. show tables, created table, etc).
             return response;
           }
         }
+
+        if (isValidTxn(txnId)) {
+          LockType lockType = LockTypeUtil.getLockTypeFromEncoding(lockChar)
+                  .orElseThrow(() -> new MetaException("Unknown lock type: " + lockChar));
+
+          if (lockType == LockType.EXCL_WRITE && blockedBy.state == LockState.ACQUIRED) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5292]: I don't think we should give up if there is already CTAS on the same table.  What happens if the 1st CTAS aborts?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 6]
```diff
@@ -13900,7 +13901,11 @@ private void addDbAndTabToOutputs(String[] qualifiedTabName, TableType type,
...
+      outputs.add(new WriteEntity(t, WriteType.CTAS));
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13905]: Why is this needed/when, we are populating outputs in CreateTableOperation as well?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 1 above)
Comment (simhadri-g) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13904]: Thanks for the review:) .
 Yes, will update this.

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java (hunk 3 above)
Comment (simhadri-g) [reply ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:27]: This was due to Intellij's auto-import. Will remove the wild card.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 5 above)
Comment (simhadri-g) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5292]: We do not know at what stage the 1st query can abort. As it is non-deterministic, we needed to make an assumption.

So when this is enabled via the conf, we will be optimistic about the outcome and assume the 1st query always succeeds. With this assumption, we can fail-early the 2nd concurrent ctas query and prevent any unnecessary move tasks and clean up that would have been associated with the 2nd query, if it was to continue until the commit stage. Also, the 2nd user will not have to wait for a long time to find out that the query failed.

But when this feature is disabled, the query will run with a pessimistic assumption that the 1st query can abort. As a result, it does not fail the 2nd query until the commit stage. This will result in a lot of overhead and clean-up associated with the failed query.  This may also make the user wait for a long time only to find out that the query failed which I think is not ideal.

This was my thought process. Would this be fine?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 6 above)
Comment (simhadri-g) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13905]: In CreateTableOperation, DDLUtils.addIfAbsentByName would populate the output only if it was missed in the SemanticAnalyzer stage. 

So I think, removing this step in CreateTableOperation and retaining it only in SemanticAnalyze would be fine.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableOperation.java [hunk 7]
```diff
@@ -99,7 +99,8 @@ public int execute() throws HiveException {
       createTableNonReplaceMode(tbl);
     }
 
-    DDLUtils.addIfAbsentByName(new WriteEntity(tbl, WriteEntity.WriteType.DDL_NO_LOCK), context);
+      DDLUtils.addIfAbsentByName(new WriteEntity(tbl, WriteEntity.WriteType.DDL_NO_LOCK), context);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableOperation.java:102]: what changed here, extra space?

Comment (simhadri-g) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableOperation.java]: Removed populating outputs in CreateTableOperation from the previous commit and retained it only in SemanticAnalyze. So, when removing the previous code, extra space has crept in. 
Will remove the extra line and space.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 5 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5292]: I don't really like that we are adding extra overhead in checkLocks method, it's already a sensitive part performance-wise. I think we should try to optimize: if it's CTAS we know that it could only be blocked by another artificial CTAS or DROP database, so no need to run expensive checkLock `BIG` query.
Also, I would expect IOW to behave similarly to CTAS, currently it doesn't fail and is executed in sequential order.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 8]
```diff
@@ -13910,7 +13911,13 @@ private void addDbAndTabToOutputs(String[] qualifiedTabName, TableType type,
     for(Map.Entry<String,String> serdeMap : storageFormat.getSerdeProps().entrySet()){
       t.setSerdeParam(serdeMap.getKey(), serdeMap.getValue());
     }
-    outputs.add(new WriteEntity(t, WriteEntity.WriteType.DDL_NO_LOCK));
+    if (tblProps != null &&
+            tblProps.get(TABLE_IS_CTAS) == "true" &&
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13915]: could we do `Boolean.parseBoolean(..)` instead of comparing with the string

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 5 above)
Comment (simhadri-g) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: I was able to optimize it by moving this check to a much earlier step, right before we call checkLocks. This prevents unnecessary insert into the TXN_COMPONENTS table and saves us at least 3 to 4 subsequent queries to the metastore.

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 9]
```diff
@@ -4667,6 +4667,9 @@ public static enum ConfVars {
     HIVE_ACID_DIRECT_INSERT_ENABLED("hive.acid.direct.insert.enabled", true,
         "Enable writing the data files directly to the table's final destination instead of the staging directory."
         + "This optimization only applies on INSERT operations on ACID tables."),
+    HIVE_ACID_NO_RENAME_CTAS_ENABLED("hive.acid.no.rename.ctas.enabled", false,
```
Comment (deniskuzZ) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:4670]: you'll probably need to rebase as CTAS change is already merged

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 10]
```diff
@@ -60,6 +60,7 @@
 import javax.sql.DataSource;
 
 import com.google.common.collect.ImmutableList;
+import jline.internal.Log;
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:63]: we should probably use `org.slf4j.Logger` as everywhere else

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 11]
```diff
@@ -3087,7 +3090,24 @@ private void insertTxnComponents(long txnid, LockRequest rqst, Connection dbConn
           String tblName = normalizeCase(lc.getTablename());
           String partName = normalizePartitionCase(lc.getPartitionname());
           OperationType opType = OperationType.fromDataOperationType(lc.getOperationType());
-
+          if (opType.getSqlConst().equals(OperationType.CTAS.getSqlConst())) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:3093]: why do you even need to create locks for CTAS with this approach? that won't work as select doesn't even use any locking and we shouldn't add one. Why can't we stick to checkLock method and only check for Excl_Write and Exclusive locks in case of CTAS operation?

Comment (simhadri-g) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: Reverted to checkLocks method in the new commit.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 12]
```diff
@@ -5297,6 +5303,28 @@ is performed on that db (e.g. show tables, created table, etc).
...
           }
         }
+
+        if (checkForConcurrentCtas && isValidTxn(txnId)) {
+          LockType lockType = LockTypeUtil.getLockTypeFromEncoding(lockChar)
+                  .orElseThrow(() -> new MetaException("Unknown lock type: " + lockChar));
+
+          if (lockType == LockType.EXCL_WRITE && blockedBy.state == LockState.ACQUIRED) {
+
+            String deleteBlockedByTxnComp = "DELETE  FROM \"TXN_COMPONENTS\" WHERE" + " \"TC_TXNID\"=" + txnId;
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5313]: no need for txn_components cleanup, txn abort should do the trick

File: ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java [hunk 13]
```diff
@@ -420,6 +420,7 @@ LockState acquireLocks(QueryPlan plan, Context ctx, String username, boolean isB
     }
     List<LockComponent> lockComponents = AcidUtils.makeLockComponents(plan.getOutputs(), plan.getInputs(),
         ctx.getOperation(), conf);
+    rqstBuilder.setCheckForConcurrentCtas(AcidUtils.isCTASOperation(lockComponents, conf));
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java:423]: could we rename it to `isExclusiveCTAS`

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 14]
```diff
@@ -4678,6 +4678,9 @@ public static enum ConfVars {
     HIVE_ACID_DIRECT_INSERT_ENABLED("hive.acid.direct.insert.enabled", true,
         "Enable writing the data files directly to the table's final destination instead of the staging directory."
         + "This optimization only applies on INSERT operations on ACID tables."),
+    HIVE_ACID_CHECK_FOR_CONCURRENT_CTAS_ENABLED("hive.acid.check.for.concurrent.ctas.enabled", false,
```
Comment (deniskuzZ) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:4681]: please rename to
````
TXN_CTAS_X_LOCK("hive.txn.xlock.ctas"
````

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 15]
```diff
@@ -3225,6 +3242,9 @@ public static String getPathSuffix(long txnId) {
     return (SOFT_DELETE_PATH_SUFFIX + String.format(DELTA_DIGITS, txnId));
   }
 
+  public static boolean isNoRenameCtasEnabled(Configuration conf) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3245]: could we please rename it to
````
isExclusiveCTASEnabled
````

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 16]
```diff
@@ -3104,6 +3112,15 @@ Seems much cleaner if each stmt is identified as a particular HiveOperation (whi
     return lockComponents;
   }
 
+  public static boolean isCTASOperation(List<LockComponent> lockComponents, HiveConf conf) {
+    boolean isCtas = false;
+    for (LockComponent lock : lockComponents) {
+      if (lock.getOperationType().name().equals(OperationType.CTAS.name()) &&
+              conf.getBoolVar(ConfVars.HIVE_ACID_CHECK_FOR_CONCURRENT_CTAS_ENABLED))
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3119]: why not use isNoRenameCtasEnabled/isExclusiveCTASEnabled helper method

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 17]
```diff
@@ -5297,6 +5303,28 @@ is performed on that db (e.g. show tables, created table, etc).
             return response;
           }
         }
+
+        if (checkForConcurrentCtas && isValidTxn(txnId)) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5307]: embed this check with above zeroWaitReadEnabled
````
if ((zeroWaitReadEnabled || isExclusiveCTAS) && isValidTxn(txnId)) {
....
if (lockType == LockType.SHARED_READ || isExclusiveCTAS) {
....
````

**Review by deniskuzZ — COMMENTED** [review]
LGTM, a few refactor-related comments

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 12 above)
Comment (simhadri-g) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5313]: No, txn abort still leaves behind entries in TXN_COMPONENTS . 
This happens even when an insert query is aborted. Looks like a bug.

Below TXN_ID = 2  was aborted. But TXN_COMPONENTS still has entry.
```
mysql> select * from TXNS;
+--------+-----------+---------------+--------------------+----------+-----------+----------------+---------------+---------------------+----------+
| TXN_ID | TXN_STATE | TXN_STARTED   | TXN_LAST_HEARTBEAT | TXN_USER | TXN_HOST  | TXN_AGENT_INFO | TXN_META_INFO | TXN_HEARTBEAT_COUNT | TXN_TYPE |
+--------+-----------+---------------+--------------------+----------+-----------+----------------+---------------+---------------------+----------+
|      0 | c         |             0 |                  0 |          |           | NULL           | NULL          |                NULL |     NULL |
|      1 | c         | 1655320638716 |      1655320638716 | hive     | localhost | NULL           | NULL          |                NULL |        0 |
|      2 | a         | 1655320644989 |      1655320664896 | hive     | localhost | NULL           | NULL          |                NULL |        0 |
|      3 | c         | 1655320688403 |      1655320688403 | hive     | localhost | NULL           | NULL          |                NULL |        0 |
|      4 | c         | 1655320719660 |      1655320719660 | hive     | localhost | NULL           | NULL          |                NULL |        0 |


mysql> select * from TXN_COMPONENTS;
+----------+-------------+----------+--------------+-------------------+------------+
| TC_TXNID | TC_DATABASE | TC_TABLE | TC_PARTITION | TC_OPERATION_TYPE | TC_WRITEID |
+----------+-------------+----------+--------------+-------------------+------------+
|        2 | default     | t1       | NULL         | i                 |          1 |
+----------+-------------+----------+--------------+-------------------+------------+
1 row in set (0.00 sec)
```

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java (hunk 16 above)
Comment (simhadri-g) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java]: Done.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 12 above)
Comment (simhadri-g) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: Realized that the cleaner will take care of this. I have removed the delete query in the recent commit.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 17 above)
Comment (simhadri-g) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: done

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 18]
```diff
@@ -4678,6 +4678,12 @@ public static enum ConfVars {
...
```
Comment (deniskuzZ) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:4681]: unused config

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 19]
```diff
@@ -5280,23 +5291,24 @@ is performed on that db (e.g. show tables, created table, etc).
...
-              .orElseThrow(() -> new MetaException("Unknown lock type: " + lockChar));
-
-          if (lockType == LockType.SHARED_READ) {
-            String cleanupQuery = "DELETE FROM \"HIVE_LOCKS\" WHERE \"HL_LOCK_EXT_ID\" = " + extLockId;
+                  .orElseThrow(() -> new MetaException("Unknown lock type: " + lockChar));
 
-            LOG.debug("Going to execute query: <" + cleanupQuery + ">");
-            stmt.executeUpdate(cleanupQuery);
+          if (lockType == LockType.SHARED_READ || isExclusiveCTAS) {
+            if (!isExclusiveCTAS) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5299]: delete from HIVE_LOCKS should be done for isExclusiveCTAS as well

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 20]
```diff
@@ -310,6 +310,11 @@ abstract class TxnHandler implements TxnStore, TxnStore.MutexAPI {
           "INNER JOIN \"TXNS\" ON \"TC_TXNID\" = \"TXN_ID\" WHERE \"TXN_STATE\" = " + TxnStatus.ABORTED +
       " GROUP BY \"TC_DATABASE\", \"TC_TABLE\", \"TC_PARTITION\" HAVING COUNT(\"TXN_ID\") > ?";
 
+  private static final String EX_CTAS_ERR_MSG =
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:313]: rename to `EXCL_CTAS_ERR_MSG`

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 21]
```diff
@@ -3086,6 +3086,13 @@ Seems much cleaner if each stmt is identified as a particular HiveOperation (whi
             output.getWriteType().name()));
         break;
 
+      case CTAS:
+        if (AcidUtils.isExclusiveCTASEnabled(conf) && AcidUtils.isTransactionalTable(t)) {
+          compBuilder.setExclWrite();
+          compBuilder.setOperationType(DataOperationType.CTAS);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3092]: set to `NO_TXN`

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 22]
```diff
@@ -3104,6 +3111,11 @@ Seems much cleaner if each stmt is identified as a particular HiveOperation (whi
     return lockComponents;
   }
 
+  public static boolean isExclusiveCTAS(List<LockComponent> lockComponents, HiveConf conf) {
+    return lockComponents.stream().anyMatch(lc -> DataOperationType.CTAS == lc.getOperationType()
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3115]: use Set<WriteEntity> outputs.getType
````
isExclusiveCTAS(work.getOutputs())
````

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/OperationType.java [hunk 23]
```diff
@@ -34,7 +34,8 @@ public enum OperationType {
   INSERT('i', DataOperationType.INSERT),
   UPDATE('u', DataOperationType.UPDATE),
   DELETE('d', DataOperationType.DELETE),
-  COMPACT('c', null);
+  COMPACT('c', null),
+  CTAS('t', DataOperationType.CTAS);
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/OperationType.java:38]: remove this

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 24]
```diff
@@ -3086,6 +3086,13 @@ Seems much cleaner if each stmt is identified as a particular HiveOperation (whi
             output.getWriteType().name()));
         break;
 
+      case CTAS:
+        if (AcidUtils.isExclusiveCTASEnabled(conf) && AcidUtils.isTransactionalTable(t)) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3090]: Could we move `AcidUtils.isTransactionalTable(t)` check inside of isExclusiveCTASEnabled?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 25]
```diff
@@ -3163,6 +3168,7 @@ private boolean shouldUpdateTxnComponent(long txnid, LockRequest rqst, LockCompo
         case INSERT:
         case UPDATE:
         case DELETE:
+        case CTAS:
```
[the remark this answers is not in the document]
Comment (simhadri-g) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:3171]: It was NO_TXN before.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java (hunk 24 above)
Comment (simhadri-g) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3090]: No, since we do not have table data at the other places where the isExclusiveCTASEnabled method is called.

File: serde/if/test/complex.thrift [hunk 26]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (deniskuzZ) [inline serde/if/test/complex.thrift:1]: nit: space

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 27]
```diff
@@ -3086,6 +3086,13 @@ Seems much cleaner if each stmt is identified as a particular HiveOperation (whi
             output.getWriteType().name()));
         break;
 
+      case CTAS:
+        if (AcidUtils.isExclusiveCTASEnabled(t, conf)) {
+          compBuilder.setExclWrite();
+          compBuilder.setOperationType(DataOperationType.NO_TXN);
+          break;
+        }
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3094]: should we continue otherwise?

Comment (simhadri-g) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3094]: Yes, it should continue if it is not a transactional table or the config is not enabled.
This was the old behavior as well.

Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3094]: i do not see `continue` keyword

**Review by deniskuzZ — APPROVED** [review]
